pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionMetaDataTopicTailer.java

```diff
@@ -77,12 +89,25 @@ public void run() {
                     if (!(th instanceof InterruptedException || th.getCause() instanceof InterruptedException)) {
                         log.warn("Encountered error when metadata tailer is not running", th);
                     }
-                    return;
                 }
             }
         }
     }
 
+    public void stopWhenNoMoreMessages() {
+        stopOnNoMessageAvailable = true;
+        readerThread.interrupt();
```

Comment:
I don't think this is the right approach.  You might be interrupting a message being processed.  The goal is the reach to the end and process all the messages.  I would just doing something similar to what I have done here:
https://github.com/apache/pulsar/pull/7237/files#diff-fb140c4ab9a86232f8d85b90cf0d3705R168

Reply:
changed